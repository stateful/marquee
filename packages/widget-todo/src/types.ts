import type { ContextProperties } from "@vscode-marquee/utils";

export interface Todo {
  body: string
  checked: boolean
  id: string
  archived: boolean
  workspaceId: string | null
  tags: string[]
  exists?: boolean
  path?: string
}

export interface State {
  todos: Todo[]
}

export interface Configuration {
  todoFilter: string
  hide: boolean
  showArchived: boolean
  autoDetect: boolean
}

export interface ContextValues extends Configuration, State {
  showAddDialog: boolean
  showEditDialog?: string
}

export interface Events {
  openAddTodoDialog: boolean
  openEditTodoDialog?: string
}

export interface Context extends Omit<ContextProperties<ContextValues>, 'setActiveWorkspaceId'> {
  _resetTodos: () => void
  _addTodo: (body: string, tags: string[], isWorkspaceTodo: boolean) => void
  _removeTodo: (id: string) => void
  _updateTodo: (todo: Todo) => void
}
