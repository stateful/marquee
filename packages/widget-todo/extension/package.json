{
  "name": "@vscode-marquee/widget-todo-extension",
  "description": "extension logic for Todo Widget",
  "main": "../build/extension.js",
  "module": "../build/extension.js",
  "types": "../build/extension.d.ts"
}
