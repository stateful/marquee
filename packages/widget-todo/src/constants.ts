import type { State, Configuration } from './types';

export const DEFAULT_STATE: State = {
  todos: []
};

export const DEFAULT_CONFIGURATION: Configuration = {
  todoFilter: '',
  hide: false,
  showArchived: false,
  autoDetect: true
};
