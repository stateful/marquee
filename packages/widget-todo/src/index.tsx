import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faTasks } from '@fortawesome/free-solid-svg-icons/faTasks'

import Todo from './Widget'

export default {
  name: 'todo',
  icon: <FontAwesomeIcon icon={faTasks} />,
  tags: ['todo', 'organization', 'productivity'],
  label: 'Todo',
  description: 'The Todo widget for simplifying priorities.',
  component: Todo,
}
export { Todo }
