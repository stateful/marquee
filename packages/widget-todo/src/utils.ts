import type { MarqueeWindow } from '@vscode-marquee/utils'
import { Todo } from './types'

declare const window: MarqueeWindow

interface FilterParams {
  globalScope?: boolean
  hide?: boolean
  todoFilter?: string
  showArchived?: boolean
}
export function filterItems (todos: Todo[], params: FilterParams) {
  let filteredItems = todos

  if (!params.globalScope) {
    filteredItems = filteredItems.filter((item) => {
      if (item['workspaceId'] === window.activeWorkspace?.id) {
        return true
      }
    })
  }

  if (!params.showArchived) {
    filteredItems = filteredItems.filter((item) => {
      if (!item.hasOwnProperty('archived') || item.archived === false) {
        return true
      }
    })
  }

  if (params.todoFilter) {
    filteredItems = filteredItems.filter((item) => {
      let inBody = item.body
        .toLowerCase()
        .indexOf(params.todoFilter!.toLowerCase()) !== -1
      let inTags = false

      if (item.tags && item.tags.length !== 0) {
        inTags =
          JSON.stringify(item.tags)
            .toLowerCase()
            .indexOf(params.todoFilter!.toLowerCase()) !== -1
      }

      if (inBody || inTags) {
        return true
      }
    })
  }

  if (params.hide) {
    filteredItems = filteredItems.filter((item) => {
      return item.checked === false
    })
  }

  return filteredItems
}
