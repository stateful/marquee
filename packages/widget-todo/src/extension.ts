import vscode from 'vscode'

import ExtensionManager from '@vscode-marquee/utils/extension'

import { DEFAULT_CONFIGURATION, DEFAULT_STATE } from './constants'
import type { Configuration, State, Todo } from './types'

const STATE_KEY = 'widgets.todo'
export const CODE_TODO = 'marquee_todo'
export const TODO = /(TODO|ToDo|Todo|todo)[:]? /

export class TodoExtensionManager extends ExtensionManager<State, Configuration> {
  constructor (context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
    super(context, channel, STATE_KEY, DEFAULT_CONFIGURATION, DEFAULT_STATE)

    const diagnostics = vscode.languages.createDiagnosticCollection('todo')
    this._refreshActiveTextEditor(diagnostics)


    this._disposables.push(
      diagnostics,

      vscode.commands.registerCommand('marquee.todo.toggle', this._toggleTodo.bind(this)),
      vscode.commands.registerCommand('marquee.todo.archive', this._archiveTodo.bind(this)),
      vscode.commands.registerCommand('marquee.todo.move', this._moveTodo.bind(this)),
      vscode.commands.registerCommand('marquee.todo.add', this._addTodo.bind(this)),
      vscode.commands.registerCommand('marquee.todo.addEmpty', () => this.emit(
        'openDialog', { event: 'openAddTodoDialog', payload: true })),
      vscode.commands.registerTextEditorCommand('marquee.todo.addEditor', this._addEditor.bind(this)),

      /**
       * diagnostics
       */
      vscode.languages.registerCodeActionsProvider('*', new TodoInfo(), {
        providedCodeActionKinds: TodoInfo.providedCodeActionKinds,
      }),
      vscode.window.onDidChangeActiveTextEditor(
        (editor) => editor && this._refreshDiagnostics(editor.document, diagnostics)),
      vscode.workspace.onDidChangeTextDocument(
        (e) => this._refreshDiagnostics(e.document, diagnostics)),
      vscode.workspace.onDidCloseTextDocument(
        (doc) => diagnostics.delete(doc.uri)),
      vscode.commands.registerCommand('marquee.refreshCodeActions',
        () => this._refreshActiveTextEditor(diagnostics))
    )
  }

  private _refreshActiveTextEditor (diagnostics: vscode.DiagnosticCollection) {
    if (vscode.window.activeTextEditor) {
      this._refreshDiagnostics(
        vscode.window.activeTextEditor.document,
        diagnostics
      )
    }
  }

  private _refreshDiagnostics (
    doc: vscode.TextDocument,
    snippetDiagnostics: vscode.DiagnosticCollection
  ): void {
    const diagnostics: vscode.Diagnostic[] = []
    const enabled = this._configuration.autoDetect === true

    for (let lineIndex = 0; lineIndex < doc.lineCount; lineIndex++) {
      const lineOfText = doc.lineAt(lineIndex)
      if (!enabled || !lineOfText.text.match(TODO)) {
        continue
      }

      const found = this.state.todos.find((todo: Todo) => {
        if (todo.path && vscode.Uri.parse(todo.path).fsPath.indexOf(doc.uri.fsPath) > -1) {
          const splt = todo.path.split(':')
          const li = parseInt(splt[splt.length - 1], 10)
          return lineIndex === li
        }

        return false
      })

      if (found) {
        continue
      }

      diagnostics.push(this._createDiagnostic(doc, lineOfText, lineIndex))
    }

    snippetDiagnostics.set(doc.uri, diagnostics)
  }

  private _createDiagnostic (
    doc: vscode.TextDocument,
    lineOfText: vscode.TextLine,
    lineIndex: number
  ): vscode.Diagnostic {
    const match = lineOfText.text.match(TODO)
    const sub = match ? match[0] : 'TODO '
    const index = lineOfText.text.indexOf(sub)
    const range = new vscode.Range(
      lineIndex,
      index,
      lineIndex,
      lineOfText.text.length
    )

    const diagnostic = new vscode.Diagnostic(
      range,
      'Add todo to Marquee?',
      vscode.DiagnosticSeverity.Information
    )
    diagnostic.code = CODE_TODO
    diagnostic.source = doc.getText(range)
    return diagnostic
  }

  /**
   * Add todo highlighted in code editor, e.g. Todo: foobar
   * see `_createDiagnostic`
   */
  private _addTodo (diagnostic: vscode.Diagnostic) {
    let body = vscode.window.activeTextEditor?.document.getText(diagnostic.range) || ''

    if (body.length < 1 || !vscode.window.activeTextEditor) {
      return
    }

    const path = `${vscode.window.activeTextEditor.document.uri.path}:${diagnostic.range.start.line}`
    body = body.replace(TODO, '').trim()

    const todo: Todo = {
      archived: false,
      body: body,
      checked: false,
      id: this.generateId(),
      tags: [],
      path,
      origin: path,
      workspaceId: this.getActiveWorkspace()?.id || null,
    }

    const newTodos = [todo].concat(this.state.todos)
    this.updateState('todos', newTodos)
    this.emit('gui.open')
    this.broadcast({ todos: newTodos })
    vscode.commands.executeCommand('marquee.refreshCodeActions')
  }

  /**
   * add todo from editor
   */
  private _addEditor (editor: vscode.TextEditor) {
    const { text, path } = this.getTextSelection(editor)

    if (text.length < 1) {
      return vscode.window.showWarningMessage('Marquee: no text selected')
    }

    const todo: Todo = {
      archived: false,
      body: text,
      checked: false,
      id: this.generateId(),
      tags: [],
      path: path,
      origin: path,
      workspaceId: this.getActiveWorkspace()?.id || null,
    }
    const newTodos = [todo].concat(this.state.todos)
    this.updateState('todos', newTodos)
    this.broadcast({ todos: newTodos })
    vscode.window.showInformationMessage(
      `Added "${text}" to your todos in Marquee`,
      'Open Marquee'
    ).then((item) => item && this.emit('gui.open'))
  }

  /**
   * toggle todo
   * @param item TreeView item that represents a todo in a tree view
   * @returns toggled todo
   */
  private _toggleTodo (item: { id: string, checked: boolean }) {
    const todoIndex = this.state.todos.findIndex((t) => t.id === item.id)

    if (todoIndex < 0) {
      return console.warn(`Couldn't find todo to toggle with id "${item.id}"`)
    }

    this.state.todos[todoIndex].checked = !item.checked
    this.updateState('todos', this.state.todos)
    this.broadcast({ todos: this.state.todos })
  }

  /**
   * archive todo
   * @param item TreeView item that represents a todo in a tree view
   * @returns (un)archived todo
   */
  private _archiveTodo (item: { id: string, archived: boolean }) {
    const todoIndex = this.state.todos.findIndex((t) => t.id === item.id)

    if (todoIndex < 0) {
      return console.warn(`Couldn't find todo to toggle with id "${item.id}"`)
    }

    this.state.todos[todoIndex].archived = !item.archived
    this.updateState('todos', this.state.todos)
    this.broadcast({ todos: this.state.todos })
  }

  /**
   * archive todo
   * @param item TreeView item that represents a todo in a tree view
   * @returns (un)archived todo
   */
  private _moveTodo (item: { id: string, archived: boolean }) {
    const awsp = this.getActiveWorkspace()

    if (!awsp) {
      return console.warn('Can\'t move todo, no workspace selected')
    }

    const todoIndex = this.state.todos.findIndex((t) => t.id === item.id)

    if (todoIndex < 0) {
      return console.warn(`Couldn't find todo to toggle with id "${item.id}"`)
    }

    this.state.todos[todoIndex].workspaceId = awsp.id
    this.updateState('todos', this.state.todos)
    this.broadcast({ todos: this.state.todos })
  }
}

class TodoInfo implements vscode.CodeActionProvider {
  public static readonly providedCodeActionKinds = [
    vscode.CodeActionKind.QuickFix,
  ]

  provideCodeActions (
    document: vscode.TextDocument,
    range: vscode.Range | vscode.Selection,
    context: vscode.CodeActionContext,
  ): vscode.CodeAction[] {
    return context.diagnostics
      .filter((diagnostic) => diagnostic.code === CODE_TODO)
      .map((diagnostic) => {
        return this.createCommandCodeAction(diagnostic)
      })
  }

  private createCommandCodeAction (
    diagnostic: vscode.Diagnostic
  ): vscode.CodeAction {
    const action = new vscode.CodeAction(
      'Marquee: Add Todo',
      vscode.CodeActionKind.QuickFix
    )
    action.command = {
      command: 'marquee.todo.add',
      title: 'Marquee: Add Todo',
      arguments: [diagnostic],
    }
    action.diagnostics = [diagnostic]
    action.isPreferred = true
    return action
  }
}

export function activate (
  context: vscode.ExtensionContext,
  channel: vscode.OutputChannel
) {
  const stateManager = new TodoExtensionManager(context, channel)

  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  }
}
export * from './types'
