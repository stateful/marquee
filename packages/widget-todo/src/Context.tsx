import React, { createContext, useState, useEffect, useContext } from 'react'
import { connect, getEventListener, MarqueeWindow, MarqueeEvents, GlobalContext } from '@vscode-marquee/utils'

import AddDialog from './dialogs/AddDialog'
import EditDialog from './dialogs/EditDialog'
import type { Todo, Context, Configuration, State, Events } from './types'

declare const window: MarqueeWindow

const TodoContext = createContext<Context>({} as Context)
const WIDGET_ID = '@vscode-marquee/todo-widget'

const TodoProvider = ({ children }: { children: React.ReactElement }) => {
  const { commit, branch } = useContext(GlobalContext)
  const eventListener = getEventListener<Events & MarqueeEvents>()
  const widgetState = getEventListener<Configuration & State>(WIDGET_ID)
  const providerValues = connect<Configuration & State>({
    ...window.marqueeStateConfiguration[WIDGET_ID].state,
    ...window.marqueeStateConfiguration[WIDGET_ID].configuration
  }, widgetState)

  const [showAddDialog, setShowAddDialog] = useState(false)
  const [showEditDialog, setShowEditDialog] = useState<string | undefined>()

  let _addTodo = (body: string, tags: string[] = [], isWorkspaceTodo = true) => {
    eventListener.emit('telemetryEvent', { eventName: 'addTodo' })
    const globalTodos: Todo[] = providerValues.todos
    const id = [...Array(8)].map(() => Math.random().toString(36)[2]).join('')
    const workspaceId = window.activeWorkspace?.id || ''
    globalTodos.unshift({
      body,
      tags,
      id,
      commit,
      branch: `${workspaceId}#${branch}`,
      checked: false,
      archived: false,
      workspaceId: isWorkspaceTodo && window.activeWorkspace
        ? workspaceId || null
        : null,
    })
    providerValues.setTodos(globalTodos)
  }

  let _removeTodo = (id: string) => {
    eventListener.emit('telemetryEvent', { eventName: 'removeTodo' })
    let globalTodos: Todo[] = providerValues.todos
    let newTodos = globalTodos.filter((todo) => todo.id !== id)
    providerValues.setTodos(newTodos)
  }

  let _updateTodo = (todo: Todo) => {
    eventListener.emit('telemetryEvent', { eventName: 'updateTodo' })
    let globalTodos: Todo[] = providerValues.todos
    let index = globalTodos.findIndex((entry) => entry.id === todo.id)
    globalTodos[index] = todo
    providerValues.setTodos(globalTodos)
  }

  let _resetTodos = () => {
    eventListener.emit('telemetryEvent', { eventName: 'resetTodo' })
    providerValues.setTodos([])
  }

  useEffect(() => {
    eventListener.on('openAddTodoDialog', setShowAddDialog)
    eventListener.on('openEditTodoDialog', setShowEditDialog)
    return () => {
      widgetState.removeAllListeners()
      eventListener.removeAllListeners()
    }
  }, [])

  return (
    <TodoContext.Provider
      value={{
        ...providerValues,

        showAddDialog,
        setShowAddDialog,
        showEditDialog,
        setShowEditDialog,

        _resetTodos,
        _addTodo,
        _removeTodo,
        _updateTodo
      }}
    >
      { showAddDialog && (
        <AddDialog close={() => setShowAddDialog(false)} />
      )}
      { showEditDialog && providerValues.todos.find((t) => showEditDialog === t.id) && (
        <EditDialog
          close={() => setShowEditDialog(undefined)}
          todo={providerValues.todos.find((t) => showEditDialog === t.id)!}
        />
      )}
      {children}
    </TodoContext.Provider>
  )
}

export default TodoContext

export { TodoProvider }
