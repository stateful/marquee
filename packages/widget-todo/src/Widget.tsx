import React, { useContext, useMemo } from "react";
import Typography from "@material-ui/core/Typography";
import AddCircle from "@material-ui/icons/AddCircleOutlined";
import { Grid, Button, IconButton } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import { List, arrayMove } from "react-movable";

import { GlobalContext, DoubleClickHelper, MarqueeWindow } from "@vscode-marquee/utils";
import wrapper, { Dragger } from "@vscode-marquee/widget";

import TodoContext, { TodoProvider } from "./Context";
import TodoPop from "./components/Pop";
import TodoInfo from "./components/Info";
import TodoFilter from "./components/Filter";
import TodoItem from "./components/Item";

declare const window: MarqueeWindow;

const useStyles = makeStyles(() => ({
  widgetTitle: {
    borderBottom: "1px solid var(--vscode-foreground)",
    padding: "8px",
  },
}));

let Todo = () => {
  const classes = useStyles();
  const {
    setTodos,
    setShowAddDialog,
    showArchived,
    todos,
    hide,
    todoFilter,
  } = useContext(TodoContext);
  const { globalScope } = useContext(GlobalContext);

  let filteredItems = useMemo(() => {
    let filteredItems = todos;

    if (!globalScope) {
      let filteredArr = filteredItems.filter((item) => {
        if (item["workspaceId"] === window.activeWorkspace?.id) {
          return true;
        }
      });
      filteredItems = filteredArr;
    }

    if (!showArchived) {
      let filteredArr = filteredItems.filter((item) => {
        if (!item.hasOwnProperty("archived") || item.archived === false) {
          return true;
        }
      });
      filteredItems = filteredArr;
    }

    if (todoFilter) {
      let filteredArr = filteredItems.filter((item) => {
        let inBody =
          item.body.toLowerCase().indexOf(todoFilter.toLowerCase()) !== -1;
        let inTags = false;

        if (item.tags && item.tags.length !== 0) {
          inTags =
            JSON.stringify(item.tags)
              .toLowerCase()
              .indexOf(todoFilter.toLowerCase()) !== -1;
        }

        if (inBody || inTags) {
          return true;
        }
      });
      filteredItems = filteredArr;
    }

    if (hide) {
      let hideArr = filteredItems.filter((item) => {
        return item.checked === false;
      });
      filteredItems = hideArr;
    }

    return filteredItems;
  }, [todos, globalScope, hide, todoFilter, showArchived]);

  return (
    <>
      <Grid item xs={1} style={{ maxWidth: "100%" }}>
        <div className={classes.widgetTitle}>
          <Grid
            container
            direction="row"
            justifyContent="space-between"
            alignItems="stretch"
            alignContent="stretch"
          >
            <Grid item>
              <Grid container direction="row">
                <Grid item>
                  <Typography variant="subtitle1">Todo</Typography>
                </Grid>
                <Grid item>
                  <TodoInfo />
                </Grid>
              </Grid>
            </Grid>

            <Grid item>
              <Grid container direction="row" spacing={1}>
                <Grid item>
                  <TodoFilter />
                </Grid>
                <Grid item>
                  <IconButton size="small" onClick={() => setShowAddDialog(true)}>
                    <AddCircle fontSize="small" />
                  </IconButton>
                </Grid>
                <Grid item>
                  <DoubleClickHelper />
                </Grid>
                <Grid item>
                  <TodoPop />
                </Grid>
                <Grid item>
                  <Dragger />
                </Grid>
              </Grid>
            </Grid>
          </Grid>
        </div>
      </Grid>
      <Grid item xs>
        <Grid
          container
          wrap="nowrap"
          direction="column"
          style={{ height: "100%" }}
        >
          <Grid item xs style={{ overflow: "auto" }}>
            {filteredItems.length === 0 && (
              <Grid
                container
                alignItems="center"
                justifyContent="center"
                style={{ height: "80%", width: "100%" }}
              >
                <Grid item>
                  <Button startIcon={<AddCircle />} variant="outlined" onClick={() => setShowAddDialog(true)}>
                    Create a todo
                  </Button>
                </Grid>
              </Grid>
            )}
            <List
              lockVertically={true}
              values={filteredItems}
              onChange={({ oldIndex, newIndex }) => {
                let constrainedArr = filteredItems;
                let firstTodo = constrainedArr[oldIndex];
                let secondTodo = constrainedArr[newIndex];

                let realFirstIndex = todos.findIndex((todo) => {
                  return todo.id === firstTodo.id;
                });
                let realSecondIndex = todos.findIndex((todo) => {
                  return todo.id === secondTodo.id;
                });

                let newTodos = arrayMove(
                  todos,
                  realFirstIndex,
                  realSecondIndex
                );
                setTodos(newTodos);
              }}
              renderList={({ children, props }) => (
                <Grid
                  container
                  direction="column"
                  style={{ padding: "8px" }}
                  {...props}
                >
                  {children}
                </Grid>
              )}
              renderItem={({ value, props, isDragged }) => (
                <TodoItem
                  key={value.id}
                  todo={value}
                  isDragged={isDragged}
                  dragProps={props}
                />
              )}
            />
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

const Widget = () => (
  <TodoProvider>
    <Todo />
  </TodoProvider>
);
export default wrapper(Widget, 'todo');
