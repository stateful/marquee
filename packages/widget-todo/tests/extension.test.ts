import vscode from 'vscode'
import { TodoExtensionManager } from '../src/extension'

const context = {
  globalState: {
    setKeysForSync: jest.fn(),
    get: jest.fn().mockReturnValue({ someRandomConfig: 'foobar' }),
    update: jest.fn()
  }
}

jest.mock('os', () => ({
  platform: () => 'some platform',
  release: () => 'some release'
}))

test('generate proper default state and configuration', () => {
  const manager = new TodoExtensionManager(context as any)
  expect(manager.configuration).toMatchSnapshot()
  expect(manager.state).toMatchSnapshot()
  expect(vscode.languages.createDiagnosticCollection).toBeCalledTimes(1)
  expect(vscode.window.onDidChangeActiveTextEditor).toBeCalledWith(
    expect.any(Function))
  expect(vscode.workspace.onDidChangeTextDocument).toBeCalledWith(
    expect.any(Function))
  expect(vscode.workspace.onDidCloseTextDocument).toBeCalledWith(
    expect.any(Function))
})

describe('_addTodo', () => {
  beforeEach(() => {
    // @ts-expect-error
    const methodMock = jest.spyOn(TodoExtensionManager.prototype, '_refreshActiveTextEditor')
    methodMock.mockImplementation() // no-op
  })

  it('with no active text editor', () => {
    const manager = new TodoExtensionManager(context as any)
    expect(manager['_addTodo']({} as any)).toBe(undefined)
  })

  it('with zero length body', () => {
    const manager = new TodoExtensionManager(context as any)

    vscode.window.activeTextEditor = {
      document: {
        getText: jest.fn().mockReturnValue(undefined)
      }
    } as any
    expect(manager['_addTodo']({ range: 123 } as any)).toBe(undefined)
    expect(vscode.window.activeTextEditor?.document.getText)
      .toBeCalledWith(123)
  })

  it('sets todo properly', () => {
    const manager = new TodoExtensionManager(context as any)
    manager['_gitProvider'] = {
      branch: 'foobar',
      commit: 'somecommit'
    } as any
    vscode.window.activeTextEditor = {
      document: {
        getText: jest.fn().mockReturnValue('ToDo: some new todo here'),
        uri: { path: '/some/path.js' }
      }
    } as any
    const diagnostic = {
      range: { start: { line: 123 } },
    } as any
    expect(manager['_addTodo'](diagnostic)).toBe(undefined)
    expect(vscode.commands.executeCommand)
      .toBeCalledWith('marquee.refreshCodeActions')
    expect(manager.emit).toBeCalledWith('gui.open')
    expect((manager.updateState as jest.Mock).mock.calls).toMatchSnapshot()
    // @ts-expect-error
    expect((manager.broadcast as jest.Mock).mock.calls).toMatchSnapshot()
  })
})
