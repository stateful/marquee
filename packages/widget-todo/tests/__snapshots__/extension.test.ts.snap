// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_addTodo sets todo properly 1`] = `
Array [
  Array [
    "todos",
    Array [
      Object {
        "archived": false,
        "body": "some new todo here",
        "checked": false,
        "id": "123457890",
        "origin": "/some/path.js:123",
        "path": "/some/path.js:123",
        "tags": Array [],
        "workspaceId": null,
      },
    ],
  ],
]
`;

exports[`_addTodo sets todo properly 2`] = `
Array [
  Array [
    Object {
      "todos": Array [
        Object {
          "archived": false,
          "body": "some new todo here",
          "checked": false,
          "id": "123457890",
          "origin": "/some/path.js:123",
          "path": "/some/path.js:123",
          "tags": Array [],
          "workspaceId": null,
        },
      ],
    },
  ],
]
`;

exports[`generate proper default state and configuration 1`] = `
Object {
  "autoDetect": true,
  "hide": false,
  "showArchived": false,
  "todoFilter": "",
}
`;

exports[`generate proper default state and configuration 2`] = `
Object {
  "todos": Array [],
}
`;
